packages/controller/src/lib/upgradeManager.ts

```diff
@@ -0,0 +1,299 @@
+import { tools } from '@iobroker/js-controller-common';
+import { valid } from 'semver';
+import { dbConnectAsync } from '@iobroker/js-controller-cli';
+import http from 'http';
+import https from 'https';
+import type { Client as ObjectsClient } from '@iobroker/db-objects-redis';
+import { setTimeout as wait } from 'timers/promises';
+
+interface UpgradeArguments {
+    /** Version of controller to upgrade too */
+    version: string;
+    /** Admin instance which triggered the upgrade */
+    adminInstance: number;
+}
+
+interface Certificates {
+    /** Public certificate */
+    certPublic: string;
+    /** Private certificate */
+    certPrivate: string;
+}
+
+interface InsecureWebServerParameters {
+    /** if https should be used for the webserver */
+    useHttps: false;
+    /** port of the web server */
+    port: number;
+}
+
+type SecureWebServerParameters = Omit<InsecureWebServerParameters, 'useHttps'> & { useHttps: true } & Certificates;
+type WebServerParameters = InsecureWebServerParameters | SecureWebServerParameters;
+
+interface GetCertificatesParams {
+    /** The objects DB */
+    objects: ObjectsClient;
+    /** Name of the public certificate */
+    certPublicName: string;
+    /** Name of the private certificate */
+    certPrivateName: string;
+}
+
+interface ServerResponse {
+    /** If the update is still running */
+    running: boolean;
+    stderr: string[];
+    stdout: string[];
+    /** if installation process succeeded */
+    success?: boolean;
+}
+
+/** Wait ms until controller is stopped */
+const STOP_TIMEOUT_MS = 3_000;
+
+class UpgradeManager {
+    /** Instance of admin to get information from */
+    private readonly adminInstance: number;
+    /** Desired controller version */
+    private readonly version: string;
+    /** Response send by webserver */
+    private readonly response: ServerResponse = {
+        running: true,
+        stderr: [],
+        stdout: []
+    };
+
+    constructor(args: UpgradeArguments) {
+        this.adminInstance = args.adminInstance;
+        this.version = args.version;
+    }
+
+    /**
+     * Parse the commands from the cli
+     */
+    static parseCliCommands(): UpgradeArguments {
+        const additionalArgs = process.argv.slice(2);
+
+        const version = additionalArgs[0];
+        const adminInstance = parseInt(additionalArgs[1]);
+
+        const isValid = !!valid(version);
+
+        if (!isValid) {
+            UpgradeManager.printUsage();
+            throw new Error('The provided version is not valid');
+        }
+
+        if (isNaN(adminInstance)) {
+            UpgradeManager.printUsage();
+            throw new Error('Please provide a valid admin instance');
+        }
+
+        return { version, adminInstance };
+    }
+
+    /**
+     * Stops the js-controller via cli call
+     */
+    async stopController(): Promise<void> {
+        await execAsync(`${tools.appNameLowerCase} stop`);
+        await wait(STOP_TIMEOUT_MS);
```

Comment:
wait only 3s?

Reply:
it was more than enough during my tests. also the stop command itself is awaited.